streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableTransformValues.java

```diff
@@ -0,0 +1,148 @@
+    }
+
+    private static <K, V, V1> V1 computeValue(final K key,
+                                              final V value,
+                                              final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer) {
+        if (value == null) {
+            return null;
+        }
+
+        return valueTransformer.transform(key, value);
+    }
+
+    private class KTableTransformValuesProcessor extends AbstractProcessor<K, Change<V>> {
+        private final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer;
+        private KeyValueStore<K, V1> store;
+        private TupleForwarder<K, V1> tupleForwarder;
+
+        private KTableTransformValuesProcessor(final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer) {
+            this.valueTransformer = Objects.requireNonNull(valueTransformer, "valueTransformer");
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(final ProcessorContext context) {
+            super.init(context);
+
+            valueTransformer.init(new ForwardingDisabledProcessorContext(context));
+
+            if (queryableName != null) {
+                final ForwardingCacheFlushListener<K, V1> flushListener = new ForwardingCacheFlushListener<>(context, sendOldValues);
+                store = (KeyValueStore<K, V1>) context.getStateStore(queryableName);
+                tupleForwarder = new TupleForwarder<>(store, context, flushListener, sendOldValues);
+            }
+        }
+
+        @Override
+        public void process(K key, Change<V> change) {
+            final V1 newValue = computeValue(key, change.newValue, valueTransformer);
+            final V1 oldValue = sendOldValues ? computeValue(key, change.oldValue, valueTransformer) : null;
+
+            if (queryableName == null) {
+                context().forward(key, new Change<>(newValue, oldValue));
+            } else {
+                store.put(key, newValue);
+                tupleForwarder.maybeForward(key, newValue, oldValue);
+            }
+        }
+    }
+
+    private class KTableTransformValuesGetter implements KTableValueGetter<K, V1> {
+
+        private final KTableValueGetter<K, V> parentGetter;
+        private final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer;
+
+        KTableTransformValuesGetter(final KTableValueGetter<K, V> parentGetter,
+                                    final ValueTransformerWithKey<? super K, ? super V, ? extends V1> valueTransformer) {
+            this.parentGetter = Objects.requireNonNull(parentGetter, "parentGetter");
+            this.valueTransformer = Objects.requireNonNull(valueTransformer, "valueTransformer");
+        }
+
+        @Override
+        public void init(final ProcessorContext context) {
+            parentGetter.init(context);
+
+            valueTransformer.init(new ForwardingDisabledProcessorContext(context));
```

Comment:
This makes me thinking: we should let least document that users should not rely on the number of generated valueTransformers in their logic, because here, we will create two instances for getter and processor, hence we'd actually get 2N instances of transformers given N tasks, etc.

Reply:
Out of interest... when are getters created and what are they used for?